docs/high/file.rst

```diff
@@ -252,44 +252,46 @@ Filenames on different systems
 
 Different operating systems (and different file systems) store filenames with
 different encodings. Additionally, in Python there are at least two different
-representations of filenames, as encoded bytes (via str on Python 2, bytes on
-Python 3) or as a unicode string (via unicode on Python 2 and str on Python 3).
-The safest bet when creating a new file is to use unicode strings on all
-systems.
+representations of filenames, as encoded ``bytes`` or as a Unicode string
+(``str`` on Python 3).
+
+h5py's high-level interfaces always return filenames as ``str``, e.g.
+:attr:`File.filename`. h5py accepts filenames as either ``str`` or ``bytes``.
+In most cases, using Unicode (``str``) paths is preferred, but there are some
+caveats.
+
+.. note::
+
+   HDF5 handles filenames as bytes (C ``char *``), and the h5py :doc:`lowlevel`
+   matches this.
 
 macOS (OSX)
 ...........
 macOS is the simplest system to deal with, it only accepts UTF-8, so using
-unicode paths will just work (and should be preferred).
+Unicode paths will just work (and should be preferred).
```

Comment:
I suspect this 'just works' only holds for Apple's own filesystems - if Macs can mount an NFS share or an ext4 hard drive, then sooner or later you'll have to contend with filenames that aren't valid UTF-8. But I suspect that's not very common, and I don't know exactly what the right answer is in that case (same as for other Unix-y systems?), so I've left this for now.

Reply:
I *think* there's some level of enforcement of this within Python's stdlib on MacOS, or at least locales play less of a role. The more Unix-y setups (with NFS/ext4 etc.) I'm not sure about.